File: test/javascript/predicates_test.js
const jstest      = require("jstest").Test
const ParseHelper = require("./parse_helper")
const Predicates   = require("../grammars/predicates")

jstest.describe("predicates", function() { with(this) {
  include(ParseHelper)

  describe("positive look-ahead", function() { with(this) {
    it("checks the first character of a word", function() { with(this) {
      assertParse(
        ["London", 10, [
          ["", 10, []],
          ["London", 10, [
            ["L", 10, []],
            ["o", 11, []],
            ["n", 12, []],
            ["d", 13, []],
            ["o", 14, []],
            ["n", 15, []]
          ]]
        ]],
        Predicates.parse("pos-name: London")
      )
    }})

    it("rejects words where the predicate does not match", function() { with(this) {
      assertThrows(SyntaxError, () => Predicates.parse("pos-name: london"))
    }})

    it("resets the cursor after matching", function() { with(this) {
      assertParse(
        ["<abc123>", 9, [
          ["", 9, []],
          ["<", 9, []],
          ["abc123", 10, [
            ["a", 10, []],
            ["b", 11, []],
            ["c", 12, []],
            ["1", 13, []],
            ["2", 14, []],
            ["3", 15, []]
          ]],
          [">", 16, []]
        ]],
        Predicates.parse("pos-seq: <abc123>")
      )
    }})

    it("uses a reference as a predicate", function() { with(this) {
      assertParse(
        ["c99", 9, [
          ["", 9, []],
          ["c99", 9, [
            ["c", 9, []],
            ["9", 10, []],
            ["9", 11, []]
          ]]
        ]],
        Predicates.parse("pos-ref: c99")
      )
    }})
  }})

  describe("negative look-ahead", function() { with(this) {
    it("checks the first character of a word", function() { with(this) {
      assertParse(
        ["word", 10, [
          ["", 10, []],
          ["word", 10, [
            ["w", 10, []],
            ["o", 11, []],
            ["r", 12, []],
            ["d", 13, []]
          ]]
        ]],
        Predicates.parse("neg-name: word")
      )
    }})

    it("rejects words where the predicate matches", function() { with(this) {
      assertThrows(SyntaxError, () => Predicates.parse("neg-name: Word"))
    }})

    it("checks for a string at the end", function() { with(this) {
      assertParse(
        ["word", 14, [
          ["word", 14, []],
          ["", 18, []]
        ]],
        Predicates.parse("neg-tail-str: word")
      )
    }})

    it("checks for a class at the end", function() { with(this) {
      assertParse(
        ["word", 16, [
          ["word", 16, []],
          ["", 20, []]
        ]],
        Predicates.parse("neg-tail-class: word")
      )
    }})

    it("checks for any char at the end", function() { with(this) {
      assertParse(
        ["word", 14, [
          ["word", 14, []],
          ["", 18, []]
        ]],
        Predicates.parse("neg-tail-any: word")
      )
    }})

    it("rejects inputs that match the negative pattern", function() { with(this) {
      assertThrows(SyntaxError, () => Predicates.parse("neg-tail-str: wordmore text"))
      assertThrows(SyntaxError, () => Predicates.parse("neg-tail-class: words"))
      assertThrows(SyntaxError, () => Predicates.parse("neg-tail-any: word "))
    }})
  }})
}})

File: test/java/PredicatesTest.java
package canopy.predicates;

import java.util.List;
import java.util.stream.Collectors;

import static org.junit.jupiter.api.Assertions.assertThrows;
import org.junit.jupiter.api.Test;

import helpers.Node;
import helpers.NodeSpec;

import test.grammars.predicates.Label;
import test.grammars.predicates.ParseError;
import test.grammars.predicates.Predicates;
import test.grammars.predicates.TreeNode;

class PositiveLookAheadTest extends ParseHelper {
    @Test
    void checksTheFirstCharacterOfAWord() throws ParseError {
        expect(Predicates.parse("pos-name: London")).toMatch(
            node("London", 10)
                .elem(node("", 10).noElems())
                .elem(node("London", 10)
                    .elem(node("L", 10).noElems())
                    .elem(node("o", 11).noElems())
                    .elem(node("n", 12).noElems())
                    .elem(node("d", 13).noElems())
                    .elem(node("o", 14).noElems())
                    .elem(node("n", 15).noElems())
                )
        );
    }

    @Test
    void rejectsWordsWhereThePredicateDoesNotMatch() {
        assertThrows(ParseError.class, () -> Predicates.parse("pos-name: london"));
    }

    @Test
    void resetsTheCursorAfterMatching() throws ParseError {
        expect(Predicates.parse("pos-seq: <abc123>")).toMatch(
            node("<abc123>", 9)
                .elem(node("", 9).noElems())
                .elem(node("<", 9).noElems())
                .elem(node("abc123", 10)
                    .elem(node("a", 10).noElems())
                    .elem(node("b", 11).noElems())
                    .elem(node("c", 12).noElems())
                    .elem(node("1", 13).noElems())
                    .elem(node("2", 14).noElems())
                    .elem(node("3", 15).noElems())
                )
                .elem(node(">", 16).noElems())
        );
    }

    @Test
    void usesAReferenceAsAPredicate() throws ParseError {
        expect(Predicates.parse("pos-ref: c99")).toMatch(
            node("c99", 9)
                .elem(node("", 9).noElems())
                .elem(node("c99", 9)
                    .elem(node("c", 9).noElems())
                    .elem(node("9", 10).noElems())
                    .elem(node("9", 11).noElems())
                )
        );
    }
}

class NegativeLookAheadTest extends ParseHelper {
    @Test
    void checksTheFirstCharacterOfAWord() throws ParseError {
        expect(Predicates.parse("neg-name: word")).toMatch(
            node("word", 10)
                .elem(node("", 10).noElems())
                .elem(node("word", 10)
                    .elem(node("w", 10).noElems())
                    .elem(node("o", 11).noElems())
                    .elem(node("r", 12).noElems())
                    .elem(node("d", 13).noElems())
                )
        );
    }

    @Test
    void rejectsWordsWhereThePredicateMatches() {
        assertThrows(ParseError.class, () -> Predicates.parse("neg-name: Word"));
    }

    @Test
    void checksForAStringAtTheEnd() throws ParseError {
        expect(Predicates.parse("neg-tail-str: word")).toMatch(
            node("word", 14)
                .elem(node("word", 14).noElems())
                .elem(node("", 18).noElems())
        );
    }

    @Test
    void checksForAClassAtTheEnd() throws ParseError {
        expect(Predicates.parse("neg-tail-class: word")).toMatch(
            node("word", 16)
                .elem(node("word", 16).noElems())
                .elem(node("", 20).noElems())
        );
    }

    @Test
    void checksForAnyCharAtTheEnd() throws ParseError {
        expect(Predicates.parse("neg-tail-any: word")).toMatch(
            node("word", 14)
                .elem(node("word", 14).noElems())
                .elem(node("", 18).noElems())
        );
    }

    @Test
    void rejectsInputsThatMatchTheNegativePattern() {
        assertThrows(ParseError.class, () -> Predicates.parse("neg-tail-str: wordmore text"));
        assertThrows(ParseError.class, () -> Predicates.parse("neg-tail-class: words"));
        assertThrows(ParseError.class, () -> Predicates.parse("neg-tail-any: word "));
    }
}

class ParseHelper {
    Node<Label> expect(TreeNode node) {
        return new NodeWrapper(node.elements.get(1));
    }

    NodeSpec<Label> node(String text, int offset) {
        return new NodeSpec<Label>(text, offset);
    }
}

class NodeWrapper implements Node<Label> {
    private TreeNode node;

    NodeWrapper(TreeNode node) {
        this.node = node;
    }

    public String text() {
        return node.text;
    }

    public int offset() {
        return node.offset;
    }

    public List<Node<Label>> elements() {
        return node.elements.stream()
            .map((elem) -> new NodeWrapper(elem))
            .collect(Collectors.toList());
    }

    public Node<Label> get(Label key) {
        return new NodeWrapper(node.get(key));
    }
}
